### ISSUE TITLE
ARM64: Questionable register allocation allow sets in asm_ahuvload

### ISSUE THREAD

**Issue by corsix** [LuaJIT/LuaJIT#1062]
There are a handful of commits at https://github.com/corsix/LuaJIT/commits/arm64-ra, the main one being 30b8966e8a, which does two things:
1. Lift `rset_exclude(gpr, idx)` to become a slightly earlier `rset_clear(gpr, idx)`, so that it also applies to the `ra_allock(as, (irt_toitype(ir->t) << 15) | 0x7fff, gpr)` a few lines down.
2. Do `if (ofs & FUSE_REG) rset_clear(gpr, ofs & 31)` at the same time, as the `ra_allock` calls here will do horrible things if they happen to choose `ofs & 31`.

I don't have a test case exercising this, as register allocator bugs are really annoying to reliably reproduce.

-----

Moving on, there's c0877c83b, which makes `asm_intmul`'s register selection slightly less constrained; it previously allowed `right` to alias `dest`, and it seems reasonable to also allow `left` to alias `dest`. It isn't _incorrect_ per se at the moment, but it stands out as odd.

Next up is ac10ea7000, which doesn't have any effect on its own, but should be completely harmless, and is a prerequisite for anyone wanting the final commit.

Finally is add6c4b4e, which I don't propose be merged (unless hidden behind a macro that isn't usually defined, in a similar style to `LUAJIT_DEBUG_RA`), but might be useful to other people: it causes the register allocator to (often) randomly choose any register from the allowed set, rather than the default (and highly desirable) behaviour of choosing a free or cheaply evictable register, which helps to exercise rare cases and flush out bugs.

**Comment by MikePall** [comment]
Sigh, I now regret that I didn't write the ARM64 JIT backend myself. So many issues over the years.
I've added the fix and the improvement. Thanks!

Good idea with the register allocation randomization!
I've reworked it to be more general and use a more 'random' algorithm. Speed is still acceptable. Anyway, this is for bug hunting and fuzz testing, not for production. Enable with: `-DLUAJIT_RANDOM_RA`

Someone please run all tests on all architectures a couple million times with this and assertions enabled. I'm sure it'll turn up something.

**Comment by Buristan** [comment]
Hi, Mike and Peter!
This randomization in RA is really nice! I hope that it will help in discovering hidden bugs.

Nevertheless, there is a problem: once we find a bug, it may be quite hard to reproduce it since we can't set randomseed for RA. I see two options here:
* We may use `math.randomseed()` with the global PRNG state (I sketched a PoC [here](https://github.com/Buristan/LuaJIT/tree/skaplun/global-math-prng)). It's quite easy, but tests may still be flaky (even for small traces) due to the use of the global PRNG state. Tested it with:

```sh
LUA_PATH="src/?.lua;;" src/luajit -Ohotloop=1 -jdump=m -e 'math.randomseed(1) for i=1,4 do end'
```

* Or we may introduce a new function (say `jit.rarandomseed()`) to set a random seed for RA  (I sketched a PoC [here](https://github.com/Buristan/LuaJIT/tree/skaplun/jit-rarandomseed)). It becomes much more bulletproof. Tested with:
```sh
LUA_PATH="src/?.lua;;" src/luajit -Ohotloop=1 -jdump=m -e 'jit.rarandomseed(0) for i=1,4 do end'
```
Unfortunately, we still have some cases with hotcount collisions and memory layout that may give us some side effects, but at least simple examples are quite stable.

Thoughts?

---
P.S. I'm not sure that this is the right ticket for my suggestions. Should I open another one instead for such situations in the future?
